File: src/main/java/com/chrisrm/idea/themes/themes/MTAbstractTheme.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 Chris Magnussen and Elior Boukhobza
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *
 */

package com.chrisrm.idea.themes.themes;

import com.chrisrm.idea.MTConfig;
import com.chrisrm.idea.MTThemeManager;
import com.chrisrm.idea.lafs.MTDarkLaf;
import com.chrisrm.idea.lafs.MTLightLaf;
import com.chrisrm.idea.themes.models.MTSerializedTheme;
import com.chrisrm.idea.themes.models.MTThemeable;
import com.chrisrm.idea.utils.MTUI;
import com.google.common.collect.Sets;
import com.intellij.openapi.util.IconLoader;
import com.intellij.ui.ColorUtil;
import com.intellij.ui.JBColor;
import com.intellij.util.IconUtil;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.plaf.ColorUIResource;
import java.awt.*;
import java.io.Serializable;
import java.util.Collections;
import java.util.Set;

import static com.chrisrm.idea.themes.lists.MTThemeResources.*;
import static com.chrisrm.idea.utils.MTColorUtils.contrastifyBackground;
import static com.chrisrm.idea.utils.MTColorUtils.contrastifyForeground;
import static com.chrisrm.idea.utils.MTUiUtils.buildResources;

@SuppressWarnings({"DuplicateStringLiteralInspection",
    "HardCodedStringLiteral",
    "SerializableHasSerializationMethods"
})
public abstract class MTAbstractTheme implements Serializable, MTThemeable, MTSerializedTheme {

  private String id;
  private String editorColorsScheme;
  private boolean dark;
  private String name;
  private String icon;
  private transient boolean isNotHighContrast;

  @SuppressWarnings({"OverridableMethodCallDuringObjectConstruction",
      "OverriddenMethodCallDuringObjectConstruction"})
  protected MTAbstractTheme() {
    init();
  }

  /**
   * Theme Builder
   */
  @SuppressWarnings("DesignForExtension")
  protected void init() {
    setId(getThemeId())
        .setIsDark(isThemeDark())
        .setEditorColorScheme(getThemeColorScheme())
        .setIcon(getThemeIcon())
        .setName(getThemeName());
  }

  /**
   * Get the theme id
   */
  @Override
  public final String toString() {
    return getId();
  }

  /**
   * Activate the theme by overriding UIManager with the theme resources and by setting the relevant Look and feel
   */
  @SuppressWarnings({"FeatureEnvy",
      "OverlyLongMethod"})
  @Override
  public final void activate() {
    final MTConfig config = MTConfig.getInstance();
    isNotHighContrast = !config.isHighContrast();
    try {
      JBColor.setDark(dark);
      IconLoader.setUseDarkIcons(dark);
      // Overridable method
      buildAllResources();

      // Apply theme accent color if said so
      if (config.isOverrideAccentColor()) {
        config.setAccentColor(ColorUtil.toHex(getAccentColorResource()));
        MTThemeManager.applyAccents(true);
      }

      if (dark) {
        UIManager.setLookAndFeel(new MTDarkLaf(this));
      } else {
        UIManager.setLookAndFeel(new MTLightLaf(this));
      }
    } catch (final UnsupportedLookAndFeelException e) {
      e.printStackTrace();
    }
  }

  /**
   * Build all resources. Overridable
   */
  @SuppressWarnings("CheckStyle")
  void buildAllResources() {
    buildResources(getBackgroundResources(), contrastifyBackground(dark, getBackgroundColorResource(), isNotHighContrast));
    buildResources(getForegroundResources(), getForegroundColorResource());
    buildResources(getTextResources(), contrastifyForeground(dark, getTextColorResource(), isNotHighContrast));
    buildResources(getSelectionBackgroundResources(), getSelectionBackgroundColorResource());
    buildResources(getSelectionForegroundResources(), getSelectionForegroundColorResource());
    buildResources(getButtonColorResources(), getButtonColorResource());
    buildResources(getSecondaryBackgroundResources(), getSecondaryBackgroundColorResource());
    buildResources(getDisabledResources(), getDisabledColorResource());
    buildResources(getContrastResources(), contrastifyBackground(dark, getContrastColorResource(), isNotHighContrast));
    buildResources(getTableSelectedResources(), getTableSelectedColorResource());
    buildResources(getSecondBorderResources(), getSecondBorderColorResource());
    buildResources(getHighlightResources(), getHighlightColorResource());

    buildResources(getTreeSelectionResources(), getTreeSelectionColorResource());
    buildResources(getNotificationsResources(), getNotificationsColorResource());

    buildNotificationsColors();
    buildFlameChartColors();
    buildFileColors();
    buildTransparentColors();
    buildTreeSelectionInactiveColors();

    UIManager.getDefaults().put("Component.grayForeground", ColorUtil.darker(getTextColorResource(), 2));
  }

  //region Getters/Setters

  /**
   * The theme name
   */
  @NotNull
  @Override
  public final String getName() {
    return name;
  }

  /**
   * Set the theme name
   */
  @Override
  public final MTAbstractTheme setName(final String name) {
    this.name = name;
    return this;
  }

  /**
   * Get the editor color scheme
   */
  @Override
  public final String getEditorColorsScheme() {
    return editorColorsScheme;
  }

  @Override
  public final MTAbstractTheme setEditorColorScheme(final String editorColorsScheme) {
    this.editorColorsScheme = editorColorsScheme;
    return this;
  }

  /**
   * The theme id
   */
  @SuppressWarnings("DesignForExtension")
  @Override
  @NotNull
  public String getId() {
    return id;
  }

  @Override
  public final MTAbstractTheme setId(final String id) {
    this.id = id;
    return this;
  }

  /**
   * Whether the theme is a dark one
   */
  @Override
  public final boolean isDark() {
    return dark;
  }

  @Override
  public final MTAbstractTheme setIsDark(final boolean dark) {
    this.dark = dark;
    return this;
  }

  @NotNull
  @Override
  public final Icon getIcon() {
    return icon != null ? IconLoader.getIcon(icon) : IconUtil.getEmptyIcon(true);
  }

  @Override
  public final MTAbstractTheme setIcon(final String icon) {
    this.icon = icon;
    return this;
  }

  /**
   * Whether the theme is a custom or external one
   */
  @SuppressWarnings("DesignForExtension")
  @Override
  public boolean isCustom() {
    return false;
  }
  //endregion

  //region Theme methods

  /**
   * Get background color custom property
   */
  @Override
  @NotNull
  public final Color getBackgroundColor() {
    return contrastifyBackground(dark, getBackgroundColorResource(), isNotHighContrast);
  }

  /**
   * Get contrast color custom property
   */
  @Override
  @NotNull
  public final Color getContrastColor() {
    return contrastifyBackground(dark, getContrastColorResource(), isNotHighContrast);
  }

  /**
   * Get foreground color custom property
   */
  @Override
  @NotNull
  public final Color getForegroundColor() {
    return contrastifyForeground(dark, getForegroundColorResource(), isNotHighContrast);
  }

  /**
   * Get background color custom property
   */
  @Override
  @NotNull
  public final Color getPrimaryColor() {
    return contrastifyForeground(dark, getTextColorResource(), isNotHighContrast);
  }

  @NotNull
  @Override
  public final Color getSelectionBackgroundColor() {
    return getSelectionBackgroundColorResource();
  }

  @NotNull
  @Override
  public final Color getSelectionForegroundColor() {
    return getSelectionForegroundColorResource();
  }

  @NotNull
  @Override
  public final Color getExcludedColor() {
    return contrastifyBackground(dark, getExcludedColorResource(), isNotHighContrast);
  }

  @NotNull
  @Override
  public final Color getNotificationsColor() {
    return getNotificationsColorResource();
  }

  @NotNull
  @Override
  public final Color getSecondBorderColor() {
    return getSecondBorderColorResource();
  }

  @NotNull
  @Override
  public final Color getDisabledColor() {
    return getDisabledColorResource();
  }

  @NotNull
  @Override
  public final Color getSecondaryBackgroundColor() {
    return getSecondaryBackgroundColorResource();
  }

  @NotNull
  @Override
  public final Color getButtonColor() {
    return getButtonColorResource();
  }

  @NotNull
  @Override
  public final Color getTableSelectedColor() {
    return getTableSelectedColorResource();
  }

  @NotNull
  @Override
  public final Color getTextColor() {
    return getTextColorResource();
  }

  @NotNull
  @Override
  public final Color getTreeSelectionColor() {
    return getTreeSelectionColorResource();
  }

  @NotNull
  @Override
  public final Color getHighlightColor() {
    return getHighlightColorResource();
  }

  @NotNull
  @Override
  public final Color getAccentColor() {
    return getAccentColorResource();
  }

  @Override
  public final void setPristine() {
    this.isNotHighContrast = true;
  }

  //endregion

  //region Other resources

  /**
   * Special treatment for notification colors
   */
  private static void buildNotificationsColors() {
    final JBColor errorColor = new JBColor(new ColorUIResource(0xef5350), new ColorUIResource(0xb71c1c));
    UIManager.put("Notification.ToolWindowError.background", errorColor);
    UIManager.put("Notification.ToolWindow.errorBackground", errorColor);
    UIManager.put("Notification.ToolWindowError.borderColor", errorColor);
    UIManager.put("Notification.ToolWindow.errorBorderColor", errorColor);

    final JBColor warnColor = new JBColor(new ColorUIResource(0xFFD54F), new ColorUIResource(0x5D4037));
    UIManager.put("Notification.ToolWindowWarning.background", warnColor);
    UIManager.put("Notification.ToolWindow.warningBackground", warnColor);
    UIManager.put("Notification.ToolWindowWarning.borderColor", warnColor);
    UIManager.put("Notification.ToolWindow.warningBorderColor", warnColor);

    final JBColor infoColor = new JBColor(new ColorUIResource(0x66BB6A), new ColorUIResource(0x1B5E20));
    UIManager.put("Notification.ToolWindowInfo.borderColor", infoColor); // deprecated
    UIManager.put("Notification.ToolWindow.infoBorderColor", infoColor); // deprecated
    UIManager.put("Notification.ToolWindow.informativeBorderColor", infoColor);

    UIManager.put("Notification.ToolWindowInfo.background", infoColor); // deprecated
    UIManager.put("Notification.ToolWindow.infoBackground", infoColor); // deprecated
    UIManager.put("Notification.ToolWindow.informativeBackground", infoColor); // deprecated
  }

  /**
   * Special treatment for flame chart colors
   */
  private static void buildFlameChartColors() {
    UIManager.put("FlameGraph.JVMBackground", MTUI.MTColor.CYAN);
    UIManager.put("FlameGraph.JVMFocusBackground", MTUI.MTColor.BLUE);
    UIManager.put("FlameGraph.JVMSearchNotMatchedBackground", MTUI.MTColor.RED);
    UIManager.put("FlameGraph.JVMFocusSearchNotMatchedBackground", MTUI.MTColor.BROWN);

    UIManager.put("FlameGraph.nativeBackground", MTUI.MTColor.YELLOW);
    UIManager.put("FlameGraph.nativeFocusBackground", MTUI.MTColor.ORANGE);
    UIManager.put("FlameGraph.nativeSearchNotMatchedBackground", MTUI.MTColor.PURPLE);
    UIManager.put("FlameGraph.nativeFocusSearchNotMatchedBackground", MTUI.MTColor.PINK);
  }

  /**
   * Special treatment for file colors
   */
  private static void buildFileColors() {
    UIManager.put("FileColor.Green", MTUI.MTColor.DARK_GREEN);
    UIManager.put("FileColor.Blue", MTUI.MTColor.DARK_BLUE);
    UIManager.put("FileColor.Yellow", MTUI.MTColor.DARK_YELLOW);
    UIManager.put("FileColor.Orange", MTUI.MTColor.DARK_ORANGE);
    UIManager.put("FileColor.Violet", MTUI.MTColor.DARK_PURPLE);
    UIManager.put("FileColor.Rose", MTUI.MTColor.DARK_PINK);
  }

  /**
   * Build transparent colors
   */
  private static void buildTransparentColors() {
    final Set<String> colors = Collections.unmodifiableSet(
        Sets.newHashSet(
            "ScrollBar.hoverTrackColor",
            "ScrollBar.trackColor",
            "ScrollBar.Mac.hoverTrackColor",
            "ScrollBar.Mac.trackColor",
            "ScrollBar.Transparent.hoverTrackColor",
            "ScrollBar.Transparent.trackColor",
            "ScrollBar.Mac.Transparent.hoverTrackColor",
            "ScrollBar.Mac.Transparent.trackColor"
        ));

    final Color transparentBackground = MTUI.Panel.getTransparentBackground();
    buildResources(colors, transparentBackground);
  }

  /**
   * Build Tree Selection Inactive Colors
   */
  private static void buildTreeSelectionInactiveColors() {
    final Set<String> colors = Collections.unmodifiableSet(
        Sets.newHashSet(
            "Tree.selectionInactiveBackground",
            "CompletionPopup.nonFocusedState",
            "CompletionPopup.nonFocusedMask",
            "List.selectionInactiveBackground",
            "Table.selectionInactiveBackground",
            "TitlePane.inactiveBackground"
        ));

    final Color transparentBackground = MTUI.Tree.getSelectionInactiveBackground();
    buildResources(colors, transparentBackground);
  }
  //endregion
}

File: src/main/java/com/mallowigi/idea/themes/themes/MTAbstractTheme.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015-2021 Elior "Mallowigi" Boukhobza
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *
 */
package com.mallowigi.idea.themes.themes

import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.util.IconLoader
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.wm.impl.IdeBackgroundUtil
import com.intellij.ui.ColorUtil
import com.intellij.ui.JBColor
import com.mallowigi.idea.MTThemeManager
import com.mallowigi.idea.config.application.MTConfig
import com.mallowigi.idea.lafs.MTDarkLaf
import com.mallowigi.idea.lafs.MTLightLaf
import com.mallowigi.idea.themes.MTAccentMode
import com.mallowigi.idea.themes.MTAccentMode.selectionColor
import com.mallowigi.idea.themes.lists.ContrastResources
import com.mallowigi.idea.themes.lists.MTThemeResources.backgroundResources
import com.mallowigi.idea.themes.lists.MTThemeResources.buttonColorResources
import com.mallowigi.idea.themes.lists.MTThemeResources.contrastResources
import com.mallowigi.idea.themes.lists.MTThemeResources.disabledResources
import com.mallowigi.idea.themes.lists.MTThemeResources.excludedResources
import com.mallowigi.idea.themes.lists.MTThemeResources.foregroundResources
import com.mallowigi.idea.themes.lists.MTThemeResources.highlightResources
import com.mallowigi.idea.themes.lists.MTThemeResources.notificationsResources
import com.mallowigi.idea.themes.lists.MTThemeResources.secondBorderResources
import com.mallowigi.idea.themes.lists.MTThemeResources.secondaryBackgroundResources
import com.mallowigi.idea.themes.lists.MTThemeResources.selectionBackgroundResources
import com.mallowigi.idea.themes.lists.MTThemeResources.selectionForegroundResources
import com.mallowigi.idea.themes.lists.MTThemeResources.selectionTransparentBackgroundResources
import com.mallowigi.idea.themes.lists.MTThemeResources.tableSelectedResources
import com.mallowigi.idea.themes.lists.MTThemeResources.textResources
import com.mallowigi.idea.themes.lists.MTThemeResources.treeSelectionResources
import com.mallowigi.idea.themes.models.MTSerializedTheme
import com.mallowigi.idea.themes.models.MTThemeable
import com.mallowigi.idea.utils.MTColorUtils.contrastifyBackground
import com.mallowigi.idea.utils.MTColorUtils.contrastifyForeground
import com.mallowigi.idea.utils.MTUI
import com.mallowigi.idea.utils.MTUI.List.selectedColor
import com.mallowigi.idea.utils.MTUI.Panel.background
import com.mallowigi.idea.utils.MTUI.Panel.editorColor
import com.mallowigi.idea.utils.MTUI.Panel.transparentBackground
import com.mallowigi.idea.utils.MTUI.Panel.transparentSelectionBackground
import com.mallowigi.idea.utils.MTUiUtils
import java.awt.Color
import java.io.FileOutputStream
import java.io.IOException
import java.io.Serializable
import java.util.Objects
import javax.swing.Icon
import javax.swing.UIManager
import javax.swing.UnsupportedLookAndFeelException
import javax.swing.plaf.ColorUIResource

/**
 * Abstract Theme for Material Themes
 *
 */
abstract class MTAbstractTheme protected constructor() : Serializable, MTThemeable, MTSerializedTheme {

  @Transient
  private var isNotHighContrast = false

  override val icon: Icon?
    get() = if (themeIcon != null) IconLoader.getIcon(themeIcon!!, MTAbstractTheme::class.java) else null

  override val isCustom: Boolean
    get() = false

  override val isNative: Boolean
    get() = false

  //region Theme Colors
  override val backgroundColor: Color
    get() = contrastifyBackground(isThemeDark, backgroundColorResource, isNotHighContrast)

  override val contrastColor: Color
    get() = contrastifyBackground(isThemeDark, contrastColorResource, isNotHighContrast)

  override val foregroundColor: Color
    get() = contrastifyForeground(isThemeDark, foregroundColorResource, isNotHighContrast)

  override val primaryColor: Color
    get() = contrastifyForeground(isThemeDark, textColorResource, isNotHighContrast)

  override val selectionBackgroundColor: Color
    get() = selectionBackgroundColorResource

  override val selectionForegroundColor: Color
    get() = selectionForegroundColorResource

  override val excludedColor: Color
    get() = contrastifyBackground(isThemeDark, excludedColorResource, isNotHighContrast)

  override val notificationsColor: Color
    get() = notificationsColorResource

  override val secondBorderColor: Color
    get() = secondBorderColorResource

  override val disabledColor: Color
    get() = disabledColorResource

  override val secondaryBackgroundColor: Color
    get() = secondaryBackgroundColorResource

  override val buttonColor: Color
    get() = buttonColorResource

  override val tableSelectedColor: Color
    get() = tableSelectedColorResource

  override val textColor: Color
    get() = textColorResource

  override val treeSelectionColor: Color
    get() = treeSelectionColorResource

  override val highlightColor: Color
    get() = highlightColorResource

  override val accentColor: Color
    get() = accentColorResource
  //endregion

  override fun toString(): String = themeId

  /**
   * Activate the theme by overriding UIManager with the theme resources and by setting the relevant Look and feel
   */
  override fun activate() {
    val config = MTConfig.getInstance()
    isNotHighContrast = !config.isHighContrast
    try {
      JBColor.setDark(isThemeDark)
      IconLoader.setUseDarkIcons(isThemeDark)
      // Overridable method
      buildAllResources()

      // Apply theme accent color if said so
      if (config.isOverrideAccentColor) {
        config.accentColor = ColorUtil.toHex(accentColorResource)
        MTThemeManager.applyAccents(true)
      }
      installBackgroundImage()

      // Set MT Look and Feel
      setLookAndFeel()
    } catch (e: UnsupportedLookAndFeelException) {
      thisLogger().error(e)
    }
  }

  /**
   * Set look and feel for the current theme
   *
   */
  @Throws(UnsupportedLookAndFeelException::class)
  protected open fun setLookAndFeel() {
    if (isThemeDark) {
      UIManager.setLookAndFeel(MTDarkLaf())
    } else {
      UIManager.setLookAndFeel(MTLightLaf())
    }
  }

  /**
   * Build all resources. Overridable
   */
  protected open fun buildAllResources() {
    MTUiUtils.buildResources(
      backgroundResources,
      contrastifyBackground(isThemeDark, backgroundColorResource, isNotHighContrast)
    )
    MTUiUtils.buildResources(foregroundResources, foregroundColorResource)
    MTUiUtils.buildResources(
      textResources,
      contrastifyForeground(isThemeDark, textColorResource, isNotHighContrast)
    )
    MTUiUtils.buildResources(selectionBackgroundResources, selectionBackgroundColorResource)
    MTUiUtils.buildResources(
      selectionTransparentBackgroundResources,
      ColorUtil.withAlpha(selectionBackgroundColorResource, 0.8)
    )
    MTUiUtils.buildResources(selectionForegroundResources, selectionForegroundColorResource)
    MTUiUtils.buildResources(buttonColorResources, buttonColorResource)
    MTUiUtils.buildResources(secondaryBackgroundResources, secondaryBackgroundColorResource)
    MTUiUtils.buildResources(disabledResources, disabledColorResource)
    MTUiUtils.buildResources(
      contrastResources,
      contrastifyBackground(isThemeDark, contrastColorResource, isNotHighContrast)
    )
    MTUiUtils.buildResources(tableSelectedResources, tableSelectedColorResource)
    MTUiUtils.buildResources(secondBorderResources, secondBorderColorResource)
    MTUiUtils.buildResources(highlightResources, highlightColorResource)
    MTUiUtils.buildResources(treeSelectionResources, treeSelectionColorResource)
    MTUiUtils.buildResources(notificationsResources, notificationsColorResource)
    MTUiUtils.buildResources(excludedResources, excludedColorResource)
    buildNotificationsColors()
    buildFlameChartColors()
    buildFileColors()
    buildTransparentColors()
    buildTabsTransparentColors()
    buildOutlineButtons()
    buildCompletionSelectionColor()
    UIManager.getDefaults()["Component.grayForeground"] = ColorUtil.darker(textColorResource, 2)
    UIManager.getDefaults()["EditorGroupsTabs.underlineHeight"] = MTConfig.getInstance().highlightThickness
  }

  //region Theme methods

  override fun setPristine() {
    isNotHighContrast = true
  }

  private fun buildTabsTransparentColors() {
    val colors = setOf(
      "EditorTabs.inactiveColoredFileBackground"
    )
    val transparentBackground = ColorUtil.withAlpha(secondaryBackgroundColorResource, 0.5)
    MTUiUtils.buildResources(colors, transparentBackground)
  }

  private fun buildOutlineButtons() {
    val colors = setOf(
      "Button.background",
      "Button.endBackground",
      "Button.startBackground"
    )
    val buttonColor = buttonColor
    val transparentBackground = if (MTConfig.getInstance().isBorderedButtons) background else buttonColor
    MTUiUtils.buildResources(colors, transparentBackground)
  }

  override fun applyContrast(apply: Boolean) {
    val contrastedColor = if (apply) contrastColor else backgroundColor
    for (resource in ContrastResources.contrastedResources) {
      UIManager.put(resource, contrastedColor)
    }
  }

  @Suppress("HardCodedStringLiteral", "NestedBlockDepth")
  private fun installBackgroundImage() {
    val currentSpec = PropertiesComponent.getInstance().getValue(IdeBackgroundUtil.FRAME_PROP)
    val oldCurrentSpec = PropertiesComponent.getInstance().getValue("old.mt.${IdeBackgroundUtil.FRAME_PROP}")
    if (!MTConfig.getInstance().isUseMaterialWallpapers) {
      removeBackgroundImage(null)
      return
    }

    try {
      val path = backgroundImage
      if (path != null) {
        val tmpImage = FileUtil.createTempFile("mtBackgroundImage", path.substring(path.lastIndexOf('.')), true)
        val resource = javaClass.classLoader.getResource(path)
        if (resource != null) {
          javaClass.classLoader.getResourceAsStream(path).use { input ->
            FileOutputStream(tmpImage).use {
              FileUtil.copy(Objects.requireNonNull(input), it)
            }
          }

          val image = tmpImage.path
          val alpha = 85.toString()
          val fill = MTUiUtils.parseEnumValue("fill", IdeBackgroundUtil.Fill.PLAIN)
          val anchor = MTUiUtils.parseEnumValue("center", IdeBackgroundUtil.Anchor.CENTER)
          val spec = StringUtil.join(arrayOf(image, alpha, fill, anchor), ",")
          PropertiesComponent.getInstance().setValue("old.mt." + IdeBackgroundUtil.FRAME_PROP, currentSpec)
          PropertiesComponent.getInstance().setValue(IdeBackgroundUtil.FRAME_PROP, spec)
          ApplicationManager.getApplication().invokeLater { IdeBackgroundUtil.repaintAllWindows() }
        } else {
          throw IllegalArgumentException("Can't load background: $path")
        }
      } else {
        removeBackgroundImage(oldCurrentSpec)
      }
    } catch (ignored: IOException) {
      thisLogger().error(ignored)
    }
  }

  //endregion

  override fun applyAccentMode() {
    val mtConfig = MTConfig.getInstance()
    val accentColor = ColorUtil.fromHex(mtConfig.accentColor)
    val darkerAccentColor = ColorUtil.darker(accentColor, 2)
    val accentColorTransparent = ColorUtil.withAlpha(accentColor, 0.5)
    val secondAccentColor = ColorUtil.fromHex(mtConfig.secondAccentColor)
    val accentMode = mtConfig.isAccentMode
    if (accentMode) {
      // Add accent resources
      MTUiUtils.buildResources(MTAccentMode.accentModeResources, accentColor)
      MTUiUtils.buildResources(MTAccentMode.darkerAccentResources, darkerAccentColor)
      MTUiUtils.buildResources(MTAccentMode.accentModeTransparentResources, accentColorTransparent)
      // Add new selection color resources
      MTUiUtils.buildResources(MTAccentMode.selectionResources, selectionColor)
      MTUiUtils.buildResources(MTAccentMode.secondAccentResources, secondAccentColor)
    }
  }

  private fun buildNotificationsColors() {
    val errorColor = JBColor(ColorUIResource(0xef9694), ColorUIResource(0xb71c1c))
    UIManager.put("Notification.ToolWindowError.background", errorColor)
    UIManager.put("Notification.ToolWindow.errorBackground", errorColor)
    UIManager.put("Notification.ToolWindowError.borderColor", errorColor)
    UIManager.put("Notification.ToolWindow.errorBorderColor", errorColor)
    val warnColor = JBColor(ColorUIResource(0xffeca0), ColorUIResource(0x5D4037))
    UIManager.put("Notification.ToolWindowWarning.background", warnColor)
    UIManager.put("Notification.ToolWindow.warningBackground", warnColor)
    UIManager.put("Notification.ToolWindowWarning.borderColor", warnColor)
    UIManager.put("Notification.ToolWindow.warningBorderColor", warnColor)
    val infoColor = JBColor(ColorUIResource(0x87bb91), ColorUIResource(0x1B5E20))
    UIManager.put("Notification.ToolWindowInfo.borderColor", infoColor)
    UIManager.put("Notification.ToolWindow.infoBorderColor", infoColor)
    UIManager.put("Notification.ToolWindow.informativeBorderColor", infoColor)
    UIManager.put("Notification.ToolWindowInfo.background", infoColor)
    UIManager.put("Notification.ToolWindow.infoBackground", infoColor)
    UIManager.put("Notification.ToolWindow.informativeBackground", infoColor)
  }

  private fun buildFlameChartColors() {
    UIManager.put("FlameGraph.JVMBackground", MTUI.MTColor.CYAN)
    UIManager.put("FlameGraph.JVMFocusBackground", MTUI.MTColor.BLUE)
    UIManager.put("FlameGraph.JVMSearchNotMatchedBackground", MTUI.MTColor.RED)
    UIManager.put("FlameGraph.JVMFocusSearchNotMatchedBackground", MTUI.MTColor.BROWN)
    UIManager.put("FlameGraph.nativeBackground", MTUI.MTColor.YELLOW)
    UIManager.put("FlameGraph.nativeFocusBackground", MTUI.MTColor.ORANGE)
    UIManager.put("FlameGraph.nativeSearchNotMatchedBackground", MTUI.MTColor.PURPLE)
    UIManager.put("FlameGraph.nativeFocusSearchNotMatchedBackground", MTUI.MTColor.PINK)
  }

  private fun buildFileColors() {
    UIManager.put("FileColor.Green", JBColor(MTUI.MTColor.GREEN, MTUI.MTColor.DARK_GREEN))
    UIManager.put("FileColor.Blue", JBColor(MTUI.MTColor.BLUE, MTUI.MTColor.DARK_BLUE))
    UIManager.put("FileColor.Yellow", JBColor(MTUI.MTColor.YELLOW, MTUI.MTColor.DARK_YELLOW))
    UIManager.put("FileColor.Orange", JBColor(MTUI.MTColor.ORANGE, MTUI.MTColor.DARK_ORANGE))
    UIManager.put("FileColor.Violet", JBColor(MTUI.MTColor.PURPLE, MTUI.MTColor.DARK_PURPLE))
    UIManager.put("FileColor.Rose", JBColor(MTUI.MTColor.RED, MTUI.MTColor.DARK_RED))
    UIManager.put("FileColor.Gray", MTUI.Label.labelInfoForeground)
  }

  private fun buildTransparentColors() {
    val colors = setOf(
      "ScrollBar.hoverTrackColor",
      "ScrollBar.trackColor",
      "ScrollBar.Mac.hoverTrackColor",
      "ScrollBar.Mac.trackColor",
      "ScrollBar.Transparent.hoverTrackColor",
      "ScrollBar.Transparent.trackColor",
      "ScrollBar.Mac.Transparent.hoverTrackColor",
      "ScrollBar.Mac.Transparent.trackColor"
    )
    val selectionColors = setOf(
      "DragAndDrop.areaBackground"
    )
    val transparentBackground = transparentBackground
    MTUiUtils.buildResources(colors, transparentBackground)
    val transparentSelectionBackground = transparentSelectionBackground
    MTUiUtils.buildResources(selectionColors, transparentSelectionBackground)
  }

  private fun buildCompletionSelectionColor() {
    val colors = setOf(
      "CompletionPopup.selectionBackground",
      "CompletionPopup.selectionInactiveBackground"
    )
    val selectedColor = selectedColor
    val invertedColor = editorColor
    val assignedColor = if (MTConfig.getInstance().isInvertedSelectionColor) invertedColor else selectedColor
    MTUiUtils.buildResources(colors, assignedColor)
  }

  private fun removeBackgroundImage(oldCurrentSpec: String?) {
    PropertiesComponent.getInstance().setValue(IdeBackgroundUtil.FRAME_PROP, oldCurrentSpec)
    PropertiesComponent.getInstance().setValue("old.mt.${IdeBackgroundUtil.FRAME_PROP}", null)
    ApplicationManager.getApplication().invokeLater { IdeBackgroundUtil.repaintAllWindows() }
  }
}
